В этом домашнем задании мы будем работать с набором данных о пассажирах парохода "Титаник", потерпевшего крушение в 1912 году (источник: соревнование [Titanic: Machine Learning from Disaster](https://www.kaggle.com/c/titanic) на платформе [Kaggle](https://www.kaggle.com)).

# Базовый уровень

*Для получения положительной оценки нужно правильно выполнить не менее 4 заданий из 6*

## Задание 1.1
Загрузите таблицу из файла `data.csv` в переменную `df` типа `pd.DataFrame`. В качестве индекса следует использовать колонку `PassengerId`.

In [1]:
import pandas as pd

In [2]:
df = pd.DataFrame()

In [3]:
df = pd.read_csv('train.csv', sep=',', index_col='PassengerId')

In [4]:
pd.set_option('display.max_rows', 15)

In [5]:
df

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
6                   0       3   
7                   0       1   
...               ...     ...   
885                 0       3   
886                 0       3   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
6                                             Moran, Mr. James    male   NaN   
7                                      McCarthy, Mr. Timothy J    male  54.0   
...                                                        ...     ...   ...   
885                                     Sutehall, Mr. Henry Jr    male  25.0   
886                       Rice, Mrs. William (Margaret Norton)  female  39.0   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
889                   Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
890                                      Behr, Mr. Karl Howell    male  26.0   
891                                        Dooley, Mr. Patrick    male  32.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S  
6                0      0            330877   8.4583   NaN        Q  
7                0      0             17463  51.8625   E46        S  
...            ...    ...               ...      ...   ...      ...  
885              0      0   SOTON/OQ 392076   7.0500   NaN        S  
886              0      5            382652  29.1250   NaN        Q  
887              0      0            211536  13.0000   NaN        S  
888              0      0            112053  30.0000   B42        S  
889              1      2        W./C. 6607  23.4500   NaN        S  
890              0      0            111369  30.0000  C148        C  
891              0      0            370376   7.7500   NaN        Q  

[891 rows x 11 columns]

## Задание 1.2
Сколько строк и столбцов в таблице?

In [6]:
df.shape

(891, 11)

Ответ: 891 строка, 11 столбцов

## Задание 1.3
Выведите на экран последние три значения столбца `Age`(возраст).

In [7]:
df.tail(3)['Age']

PassengerId
889     NaN
890    26.0
891    32.0
Name: Age, dtype: float64

## Задание 1.4

Какова доля выживших (поле `Survived`)?

In [8]:
df[df.Survived == 1]['Name'].count() * 1.0 / df['Name'].count()

0.38383838383838381

## Задание 1.5
Выведите на экран таблицу, отсортированную по имени по алфавиту (поле `Name`).

In [9]:
df.sort_values(by='Name')

Survived  Pclass                                          Name  \
PassengerId                                                                   
846                 0       3                           Abbing, Mr. Anthony   
747                 0       3                   Abbott, Mr. Rossmore Edward   
280                 1       3              Abbott, Mrs. Stanton (Rosa Hunt)   
309                 0       2                           Abelson, Mr. Samuel   
875                 1       2         Abelson, Mrs. Samuel (Hannah Wizosky)   
366                 0       3                Adahl, Mr. Mauritz Nils Martin   
402                 0       3                               Adams, Mr. John   
...               ...     ...                                           ...   
423                 0       3                            Zimmerman, Mr. Leo   
560                 1       3  de Messemaeker, Mrs. Guillaume Joseph (Emma)   
287                 1       3                       de Mulder, Mr. Theodore   
283                 0       3                     de Pelsmaeker, Mr. Alfons   
362                 0       2                     del Carlo, Mr. Sebastiano   
154                 0       3               van Billiard, Mr. Austin Blyler   
869                 0       3                   van Melkebeke, Mr. Philemon   

                Sex   Age  SibSp  Parch         Ticket     Fare Cabin Embarked  
PassengerId                                                                     
846            male  42.0      0      0      C.A. 5547   7.5500   NaN        S  
747            male  16.0      1      1      C.A. 2673  20.2500   NaN        S  
280          female  35.0      1      1      C.A. 2673  20.2500   NaN        S  
309            male  30.0      1      0      P/PP 3381  24.0000   NaN        C  
875          female  28.0      1      0      P/PP 3381  24.0000   NaN        C  
366            male  30.0      0      0         C 7076   7.2500   NaN        S  
402            male  26.0      0      0         341826   8.0500   NaN        S  
...             ...   ...    ...    ...            ...      ...   ...      ...  
423            male  29.0      0      0         315082   7.8750   NaN        S  
560          female  36.0      1      0         345572  17.4000   NaN        S  
287            male  30.0      0      0         345774   9.5000   NaN        S  
283            male  16.0      0      0         345778   9.5000   NaN        S  
362            male  29.0      1      0  SC/PARIS 2167  27.7208   NaN        C  
154            male  40.5      0      2       A/5. 851  14.5000   NaN        S  
869            male   NaN      0      0         345777   9.5000   NaN        S  

[891 rows x 11 columns]

## Задание 1.6
Каковы доли женщин и мужчин среди пассажиров (поле `Sex`)?

In [10]:
df.groupby('Sex')['Name'].count() * 1.0 / df['Name'].count()

Sex
female    0.352413
male      0.647587
Name: Name, dtype: float64

# Продвинутый уровень

*Для получения положительной оценки нужно правильно выполнить не менее 3 заданий из 5*

## Задание 2.1
Каков процент женщин и мужчин среди пассажиров? (Hint: *процент – это доля, умноженная на 100.*)

In [11]:
df.groupby('Sex')['Name'].count() * 100.0 / df['Name'].count()

Sex
female    35.241302
male      64.758698
Name: Name, dtype: float64

## Задание 2.2
Какова доля совершеннолетних пассажиров (18 лет и старше)?

In [12]:
df[df.Age >= 18]['Name'].count() * 1.0 / df['Name'].count()

0.67452300785634123

## Задание 2.3
Замените коды пассажирских классов (поле `Pclass`) их названиями (1 = Upper, 2 = Middle, 3 = Lower). Результат выведите на экран.

In [13]:
def price_class(price):
    return 'Upper' if price == 1 else 'Middle' if price == 2 else 'Lower'

In [14]:
df['Pclass'] = df.Pclass.apply(price_class)

In [15]:
df

Survived  Pclass  \
PassengerId                     
1                   0   Lower   
2                   1   Upper   
3                   1   Lower   
4                   1   Upper   
5                   0   Lower   
6                   0   Lower   
7                   0   Upper   
...               ...     ...   
885                 0   Lower   
886                 0   Lower   
887                 0  Middle   
888                 1   Upper   
889                 0   Lower   
890                 1   Upper   
891                 0   Lower   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
6                                             Moran, Mr. James    male   NaN   
7                                      McCarthy, Mr. Timothy J    male  54.0   
...                                                        ...     ...   ...   
885                                     Sutehall, Mr. Henry Jr    male  25.0   
886                       Rice, Mrs. William (Margaret Norton)  female  39.0   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
889                   Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
890                                      Behr, Mr. Karl Howell    male  26.0   
891                                        Dooley, Mr. Patrick    male  32.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S  
6                0      0            330877   8.4583   NaN        Q  
7                0      0             17463  51.8625   E46        S  
...            ...    ...               ...      ...   ...      ...  
885              0      0   SOTON/OQ 392076   7.0500   NaN        S  
886              0      5            382652  29.1250   NaN        Q  
887              0      0            211536  13.0000   NaN        S  
888              0      0            112053  30.0000   B42        S  
889              1      2        W./C. 6607  23.4500   NaN        S  
890              0      0            111369  30.0000  C148        C  
891              0      0            370376   7.7500   NaN        Q  

[891 rows x 11 columns]

## Задача 2.4
Какова доля выживших среди женщин? Среди мужчин?

In [16]:
df.groupby('Sex')['Survived'].value_counts(normalize=True)

Sex     Survived
female  1           0.742038
        0           0.257962
male    0           0.811092
        1           0.188908
Name: Survived, dtype: float64

## Задача 2.5
Сколько всего пропущенных значений в таблице?

# Сложный уровень

*Для получения положительной оценки нужно правильно выполнить не менее 2 заданий из 3*

## Задание 3.1
Добавьте в таблицу поле `Surname` – фамилию пассажира.

In [17]:
def family(line):
    split_line = line.split(', ')
    return split_line[0]

In [18]:
df['Surname'] = df.Name.apply(family)

In [19]:
df

Survived  Pclass  \
PassengerId                     
1                   0   Lower   
2                   1   Upper   
3                   1   Lower   
4                   1   Upper   
5                   0   Lower   
6                   0   Lower   
7                   0   Upper   
...               ...     ...   
885                 0   Lower   
886                 0   Lower   
887                 0  Middle   
888                 1   Upper   
889                 0   Lower   
890                 1   Upper   
891                 0   Lower   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
6                                             Moran, Mr. James    male   NaN   
7                                      McCarthy, Mr. Timothy J    male  54.0   
...                                                        ...     ...   ...   
885                                     Sutehall, Mr. Henry Jr    male  25.0   
886                       Rice, Mrs. William (Margaret Norton)  female  39.0   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
889                   Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
890                                      Behr, Mr. Karl Howell    male  26.0   
891                                        Dooley, Mr. Patrick    male  32.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked    Surname  
PassengerId                                                                     
1                1      0         A/5 21171   7.2500   NaN        S     Braund  
2                1      0          PC 17599  71.2833   C85        C    Cumings  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  Heikkinen  
4                1      0            113803  53.1000  C123        S   Futrelle  
5                0      0            373450   8.0500   NaN        S      Allen  
6                0      0            330877   8.4583   NaN        Q      Moran  
7                0      0             17463  51.8625   E46        S   McCarthy  
...            ...    ...               ...      ...   ...      ...        ...  
885              0      0   SOTON/OQ 392076   7.0500   NaN        S   Sutehall  
886              0      5            382652  29.1250   NaN        Q       Rice  
887              0      0            211536  13.0000   NaN        S   Montvila  
888              0      0            112053  30.0000   B42        S     Graham  
889              1      2        W./C. 6607  23.4500   NaN        S   Johnston  
890              0      0            111369  30.0000  C148        C       Behr  
891              0      0            370376   7.7500   NaN        Q     Dooley  

[891 rows x 12 columns]

## Задание 3.2
Найдите всех пассажиров с аномальной стоимостью билета (поле `Fare`). Будем считать стоимость аномальной, если она отличается от среднего более, чем на 5 стандартных отклонений. Выведите соответствующие строки таблицы на экран.

## Задание 3.3
Выведите на экран столбцы, в которых бывают пропущенные значения. (Hint: *Здесь может помочь метод* `.any()`*, а также индексация переменной* `df.columns`).